keadm/cmd/keadm/app/cmd/edge/batchjoin.go

```diff
@@ -0,0 +1,179 @@
+	errCh := make(chan error, len(config.Nodes)) // Used to collect errors
+
+	for _, node := range config.Nodes {
+		wg.Add(1) // Each node corresponds to a goroutine
+		go func(n Node) {
+			defer wg.Done() // Ensure Done is called when the goroutine finishes
+			fmt.Printf("Processing node %s \n", n.IP)
+			// Perform download and copy of keadm
+			if err := downloadAndCopyKeadmToNode(ctx, &n); err != nil {
+				errCh <- err // If an error occurs, send it to the error channel
+				fmt.Printf("Processing node %s  downloadAndCopyKeadmToNode error \n", n.IP)
+				return
+			}
+			// Execute the keadm join command
+			if err := executeJoinCommand(ctx, &n, &config.KeadmParams); err != nil {
+				fmt.Printf("Processing node %s  executeJoinCommand error \n", n.IP)
+				errCh <- err // Send to the error channel when an error occurs
+				return
+			}
+			fmt.Printf("Node %s joined successfully\n", n.IP)
+		}(node)
+	}
+
+	// Wait for all goroutines to complete
+	wg.Wait()
+	close(errCh)
+
+	// Check for errors
+	for err := range errCh {
+		if err != nil {
+			return err // 如果有任意一个错误，返回第一个错误
+		}
+	}
+
+	return nil // All nodes processed successfully
+}
+
+func downloadAndCopyKeadmToNode(ctx context.Context, node *Node) error {
+	// Download keadm from GitHub Release
+	releaseURL := fmt.Sprintf("https://github.com/kubeedge/kubeedge/releases/download/%s/keadm-%s.tar.gz", node.KubeedgeVersion, node.Architecture)
+	fmt.Println("releaseURL:", releaseURL)
+	// SSH connect to the node
+	client, err := createSSHClient(node.Username, node.IP, node.Password)
+	if err != nil {
+		return fmt.Errorf("failed to connect to node %s: %v", node.IP, err)
+	}
+	defer client.Close()
+
+	// Execute remote command to download keadm
+	session, err := client.NewSession()
+	if err != nil {
+		return fmt.Errorf("failed to create new SSH session: %v", err)
+	}
+	defer session.Close()
+	fileName := fmt.Sprintf("keadm-%s.tar.gz", node.Architecture)
+	cmd := fmt.Sprintf("curl -L %s -o /tmp/%s", releaseURL, fileName)
+	if err := session.Run(cmd); err != nil {
+		return fmt.Errorf("failed to download keadm on node %s: %v", node.IP, err)
+	}
+	return nil
+}
+
+func executeJoinCommand(ctx context.Context, node *Node, params *KeadmParams) error {
+	// SSH connect to the node
+	client, err := createSSHClient(node.Username, node.IP, node.Password)
+	if err != nil {
+		return fmt.Errorf("failed to connect to node %s: %v", node.IP, err)
+	}
+	defer client.Close()
+
+	// Execute the keadm join command
+	session, err := client.NewSession()
+	if err != nil {
+		return fmt.Errorf("failed to create new SSH session: %v", err)
+	}
+	defer session.Close()
+	fileName := fmt.Sprintf("keadm-%s.tar.gz", node.Architecture)
+	dirName := fmt.Sprintf("keadm-%s", node.Architecture)
+	cmd := fmt.Sprintf("cd /tmp && tar xzf %s && cd %s/keadm && ./keadm join --cloudcore-ipport=%s --token=%s --kubeedge-version=%s --cgroupdriver systemd --runtimetype=docker",
```

Comment:
I am not sure if we can expect `--runtimetype=docker`, probably this is something that user want to configure? for example, we do use `cotnainerd` but `dockerd`...

Reply:
Yes, the `--runtiemtype` flag was deprecated, we need to set the `--remote-runtime-endpoint` flag for different container runtime.

Reply:
> I am not sure if we can expect `--runtimetype=docker`, probably this is something that user want to configure? for example, we do use `cotnainerd` but `dockerd`...



> Yes, the `--runtiemtype` flag was deprecated, we need to set the `--remote-runtime-endpoint` flag for different container runtime.

Thank you for your valuable suggestions, your suggestions are very valuable and I will make relevant improvements.